R/Spatial.R

```diff
@@ -202,4 +202,156 @@ RLE_Plot <- function(seuratObj, assay = "RNA", sampleIdentifier = "SegmentDispla
+#' 
+#' #define plotting layout
+#' layout <- "
+#' #AAAA#
+#' BBBCCC
+#' "
+#' 
+#' #Plot results
+#' ggplot(metadata %>% filter(fov == 1), 
+#' aes(x = x_FOV_px, y = y_FOV_px, color = simple_cellType)) + 
+#' geom_point() + 
+#' egg::theme_article() + 
+#' ggtitle('Cell Type assignment')
+#' ggplot(metadata %>% filter(fov == 1), aes(x = x_FOV_px, y = y_FOV_px, color = factor(Local_BCF))) + 
+#' geom_point() + 
+#' egg::theme_article() +  
+#' ggtitle('Specific substructure cell assignment') + 
+#' ggplot(metadata %>% filter(fov == 1), 
+#' aes(x = x_FOV_px, y = y_FOV_px, color = Within_Local_BCF)) + 
+#' geom_point() + 
+#' egg::theme_article() + 
+#' ggtitle('Non-specific substructure cell assignment') + 
+#' plot_layout(design = layout, guides = "collect")
+#' }
+#' @export
+DetectCellStructuresBasedOnCellType <- function(seuratObjectMetadata, 
+                                                cellTypeField = "cell_type", 
+                                                minimumClusterSizeCoefficient = 0.05,
+                                                fovField = "fov",
+                                                fovWhitelist = NULL,
+                                                cellTypeWhiteList = c("Bcell", "B_cell", "B.cell"),
+                                                xCoordinateField = "x_FOV_px", 
+                                                yCoordinateField = "y_FOV_px", 
+                                                substructureMetaDataFieldName = "BCF",
+                                                summarizeLocalResults = TRUE
+){
+  #check that inputs exist and summarizeLocalResults is a boolean. 
+  if (!(all(c(cellTypeField, fovField, xCoordinateField, yCoordinateField) %in% colnames(seuratObjectMetadata)))) {
+    missingColumn <- c(cellTypeField, fovField, xCoordinateField, yCoordinateField)[which(!(c(cellTypeField, fovField, xCoordinateField, yCoordinateField) %in% colnames(seuratObjectMetadata)))]
+    stop(paste0("Error: ", missingColumn , " was not found in the supplied metadata's column names. Please add ", missingColumn, " to the metadata."))
+  } else if (!rlang::is_bool(summarizeLocalResults)) {
+    stop("Please set summarizeLocalResults to either TRUE or FALSE.")
+  }
+  
+  #determine FOVs to loop over
+  if (is.null(fovWhitelist)) {
+    fovs <- unique(seuratObjectMetadata[,fovField])
+  } else if (!fovWhitelist %in% unique(seuratObjectMetadata[,fovField])){
+    stop(paste0("The FOVs listed in whitelist are not detected in fovField: ", fovField, ". Please ensure the fovField is formatted as you expect (specifically: ensure integer vs character typing)."))
+  } else {
+    fovs <- fovWhitelist
+  }
+  
+  #Detect substructures within each FOV
+  for (fov in fovs) {
+    fov_seuratObjectMetadata <- seuratObjectMetadata[seuratObjectMetadata[fovField] == fov,]
+    #escape special characters in cell types
+    escapedCellTypes <- gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\\1", cellTypeWhiteList)
```

Comment:
being picky, there is no reason to repeatedly do this within a loop

Reply:
true. fixed.